libsolidity/analysis/TypeChecker.cpp

```diff
@@ -299,33 +299,46 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
 		if (!_function.isConstructor() && !_function.isFallback() && !_function.isPartOfExternalInterface())
 			m_errorReporter.typeError(_function.location(), "Internal functions cannot be payable.");
 	}
-	for (ASTPointer<VariableDeclaration> const& var: _function.parameters() + _function.returnParameters())
-	{
-		if (type(*var)->category() == Type::Category::Mapping)
+	auto checkArgumentAndReturnParameter = [&](VariableDeclaration const& var) {
+		if (type(var)->category() == Type::Category::Mapping)
 		{
-			if (!type(*var)->dataStoredIn(DataLocation::Storage))
-				m_errorReporter.typeError(var->location(), "Mapping types can only have a data location of \"storage\"." );
+			if (!type(var)->dataStoredIn(DataLocation::Storage))
+				m_errorReporter.typeError(var.location(), "Mapping types can only have a data location of \"storage\"." );
 			else if (!isLibraryFunction && _function.isPublic())
-				m_errorReporter.typeError(var->location(), "Mapping types for parameters or return variables can only be used in internal or library functions.");
+				m_errorReporter.typeError(var.location(), "Mapping types for parameters or return variables can only be used in internal or library functions.");
 		}
 		else
 		{
-			if (!type(*var)->canLiveOutsideStorage() && _function.isPublic())
-				m_errorReporter.typeError(var->location(), "Type is required to live outside storage.");
-			if (_function.isPublic() && !(type(*var)->interfaceType(isLibraryFunction)))
-				m_errorReporter.fatalTypeError(var->location(), "Internal or recursive type is not allowed for public or external functions.");
+			if (!type(var)->canLiveOutsideStorage() && _function.isPublic())
+				m_errorReporter.typeError(var.location(), "Type is required to live outside storage.");
+			if (_function.isPublic() && !(type(var)->interfaceType(isLibraryFunction)))
+				m_errorReporter.fatalTypeError(var.location(), "Internal or recursive type is not allowed for public or external functions.");
 		}
 		if (
 			_function.isPublic() &&
 			!_function.sourceUnit().annotation().experimentalFeatures.count(ExperimentalFeature::ABIEncoderV2) &&
-			!typeSupportedByOldABIEncoder(*type(*var))
+			!typeSupportedByOldABIEncoder(*type(var))
 		)
 			m_errorReporter.typeError(
-				var->location(),
+				var.location(),
 				"This type is only supported in the new experimental ABI encoder. "
 				"Use \"pragma experimental ABIEncoderV2;\" to enable the feature."
 			);
-
+	};
+	for (ASTPointer<VariableDeclaration> const& var: _function.parameters())
+	{
+		if (
+			!m_scope->isInterface() &&
+			type(*var)->category() == Type::Category::Struct &&
+			type(*var)->dataStoredIn(DataLocation::CallData)
```

Comment:
This will be false for the return parameter - so it can actually just be pulled into the main loop

Reply:
Yes, but I was about to split this loop before and I think it *should* be split - it performs an unnecessary vector copy and I'm actually surprised that so far there's no difference between the checks for arguments and return parameters - I think that'll happen sooner or later anyways. And if it is split, then this check should be here.

Reply:
Since it's not necessarily *required* for this PR, though, I'm fine with *not* doing it here.